imap/jmap_mailbox.c

```diff
@@ -3395,82 +3441,343 @@ static void _mboxset_state_conflict_cb(const char *id, void *data, void *rock)
     }
 }
 
-static int _mboxset_state_is_valid(const char *account_id, struct mboxset_ops *ops)
+static void _mboxset_state_update_mboxtree(struct mboxset_state *state,
+                                           struct mboxset *set __attribute__((unused)),
+                                           struct mboxset_ops *ops)
 {
-    /* Create in-memory mailbox tree */
-    struct mboxset_state *state = xzmalloc(sizeof(struct mboxset_state));
-    state->account_id = account_id;
-
-    hash_table *id_by_imapname = xzmalloc(sizeof(struct hash_table));
-    construct_hash_table(id_by_imapname, 1024, 0);
-    state->id_by_imapname = id_by_imapname;
-
-    hash_table *entry_by_id = xzmalloc(sizeof(struct hash_table));
-    construct_hash_table(entry_by_id, 1024, 0);
-    state->entry_by_id = entry_by_id;
-
-    hash_table *siblings_by_parent_id = xzmalloc(sizeof(struct hash_table));
-    construct_hash_table(siblings_by_parent_id, 1024, 0);
-    state->siblings_by_parent_id = siblings_by_parent_id;
-
-    /* Map current mailboxes by IMAP name to id */
-    mboxlist_usermboxtree(account_id, NULL, _mboxset_state_mboxlist_cb, state,
-                          MBOXTREE_INTERMEDIATES);
-    char *inboxname = mboxname_user_mbox(account_id, NULL);
-    const char *inbox_id = hash_lookup(inboxname, id_by_imapname);
-    free(inboxname);
-
-    /* Create entries for current mailboxes */
-    hash_enumerate(id_by_imapname, _mboxset_state_mkentry_cb, state);
-
-    int i;
-    int is_valid = 0;
+    struct buf buf = BUF_INITIALIZER;
+    char *inboxname = mboxname_user_mbox(state->req->accountid, NULL);
+    const char *inbox_id = hash_lookup(inboxname, state->id_by_imapname);
 
     /* Apply create and update */
+    int i;
     for (i = 0; i < ptrarray_size(ops->put); i++) {
         struct mboxset_args *args = ptrarray_nth(ops->put, i);
-        struct buf buf = BUF_INITIALIZER;
         if (args->creation_id) {
+            /* Create entry for in-memory mailbox tree */
             struct mboxset_entry *entry = xzmalloc(sizeof(struct mboxset_entry));
             entry->parent_id = xstrdup(args->parent_id ? args->parent_id : inbox_id);
             entry->name = xstrdup(args->name);
             buf_putc(&buf, '#');
             buf_appendcstr(&buf, args->creation_id);
-            hash_insert(buf_cstring(&buf), entry, entry_by_id);
+            hash_insert(buf_cstring(&buf), entry, state->entry_by_id);
             buf_reset(&buf);
         }
         else {
-            struct mboxset_entry *entry = hash_lookup(args->mbox_id, entry_by_id);
+            /* Update entry in in-memory mailbox tree */
+            struct mboxset_entry *entry = hash_lookup(args->mbox_id, state->entry_by_id);
             if (!entry) goto done;
             if (args->name) {
                 free(entry->name);
                 entry->name = xstrdup(args->name);
             }
             if (args->parent_id) {
-                free(entry->parent_id);
```

Comment:
can this ever be called twice times?  If so we want `free(entry->old_parent_id)` before setting it.

Reply:
it should not, but it doesn't hurt to free old_parent_id